### ISSUE TITLE
Feature Request: Rust 1.51 "const generics" (if and when they become possible)

### ISSUE THREAD

**Issue by ProfRon** [google/evcxr#165]
Rust 1.51.0 is now out, and it includes a new feature called "Const Generics" (well, not entirely new - but see the [rust blog](https://blog.rust-lang.org/2021/03/25/Rust-1.51.0.html#const-generics-mvp) for all the details).  This feature does not appear to work with evcxr-jupiter.  

Here's a minimal test case -- I've made a Jupyter cell with this code (taken, slightly changed, from the blog entry linked above):
```rust
pub struct TestArray<T, const LENGTH: usize> {
    //                 ^^^^^^^^^^^^^^^^^^^ Const generic definition.
    pub list: [T; LENGTH]
    //        ^^^^^^ We use it here.
}

let test_array = TestArray::<u8,42>{ list: [0u8; 42] };

assert_eq!( test_array.list.len(), 42 );

test_array.list
```

I get the following output:

> pub struct TestArray<T, const LENGTH: usize> {
>                      ^ 
> pub struct TestArray<T, const LENGTH: usize> {
>                               ^^^^^^ 
> pub struct TestArray<T, const LENGTH: usize> {
>            ^^^^^^^^^ 
> struct defined here, with 2 generic parameters: `T`, `LENGTH`

However, when I put that code in the crate I'm working on (as a test case), it compiles and runs just fine!

I tried reinstalling evcxr-jupyter to see if that would help it use the latest rust version, with no luck.  I suppose I may have broken something in the process, so I'm very curious if it's just my environment. Does that code work for anyone else?

I haven't tried cloning the repository to make sure I have the bleeding edge version. Would it be helpful if I did, or should I just be patient for a new version to come out (and use "cargo install")?

Thanks in advance!

**Comment by bjorn3** [comment]
Evcxr used to use rustc to get the types for variables, but at some point rustc started to only show the last part of the type name if it is unambiguous instead of showing the full path. For this reason evcxr has to switch to rust-analyzer to get the types. Rust-analyzer doesn't support const generics and will likely not for a long time as it doesn't have a const evaluator as necessary to determine the value of the const generics.

**Comment by ProfRon** [comment]
OK. Thanks for the explanation. I've changed the title of this item to make it clear that it would be "nice to have" (but is apparently not possible at the moment).

I was looking at your source code yesterday (after posting this issue) and had begun to suspect that the problem is with rust-analyzer. I found the crate but no repository, so I'm not sure where I could post a feature request for this for rust-analyzer. 

I've been using evxcr-jupyter together with a "normal IDE", working up my thoughts first in Jupyter and then coding them into a "normal" crate, but now I'm stuck, having put const generics into my crate.  This is rather disappointing, but I guess I have no one to blame but myself. Now I have to decide whether to rip out the const generics again, or just do without jupyter for the time being.  

Bother. In any case, thanks for providing evxcr, and for the help it's been so far. Looking forward to using it again after it gets const generics (or I decide to live without them).

**Comment by davidlattimore** [comment]
As a workaround, does it work if you put explicit type annotations on your variables?

**Comment by bjorn3** [comment]
> In any case, thanks for providing evxcr, and for the help it's been so far.

@davidlattimore is actually the author. :)

**Comment by ProfRon** [comment]
> As a workaround, does it work if you put explicit type annotations on your variables?

If this is what you mean (for example):
```rust
pub struct TestArray<T, const LENGTH: usize> {
    pub list: [T; LENGTH]
}

let test_array : TestArray<u8,42> = TestArray{ list: [0u8; 42] };

assert_eq!( test_array.list.len(), 42 );

test_array.list
```

then no, I get the same output as before (see my original post).

(By the way, it's much the same if you leave out "T" and only use one parameter to the generic (if that parameter is a const)).

(Another "By the way" -- sorry for confusing @davidlattimore and @bjorn3, and thank *you* @davidlattimore, for all your work on this project).

**Comment by ProfRon** [comment]
Sorry to keep flogging a dead (or at least comatose) horse, but here's _another_ workaround that doesn't work (around):

If I:
1.  define a "type alias" in an "extern crate" to be a generic struct, e.g. "pub type TestArraBytey42 = TestArray<u8,42>",
2.  then try using the type alias  in evcxr-jupyter, e.g. use "TestArraBytey42" where I would have used "TestArray<u8,42>"...
    it still doesn't work.
 
Says "missing generics for struct...". 

So apparently "rust-analyzer" can and does get the complete type out of an external crate! Rust continues to amaze me...

**Comment by davidlattimore** [comment]
No problem @ProfRon, evcxr has contributions from many besides myself. @bjorn3 has contributed by helping to answer many questions, which is greatly appreciated.

It looks like the relevant issue in rust-analyzer is probably https://github.com/rust-analyzer/rust-analyzer/issues/7432

You can see the effect in rust-analyzer  if you write some code like:

```rust
struct Foo<T, const I: usize> {
    data: [T; I],
}

fn main() {
    let f = Foo { data: [9u8, 15]};
    bar(f);
    println!("{}", f.data[1]);
}
```

Then (in vscode) press control-. on `bar` and select "Generate `bar` function". It then generates the function:

```rust
fn bar(f: Foo<u8>) -> () {
    todo!()
}
```

i.e. it's missing the const parameter in the function argument.

It's unfortunate that explicit type annotations don't help here. It's possible that part might be fixable on the evcxr side of things with a little work

**Comment by JohnScience** [comment]
As a temporary solution I've created [`execute_evcxr`](https://github.com/JohnScience/execute_evcxr). It's in the process of documentation but already works more or less as expected. It supports only evcxr-flavored Rust syntax which follows pattern

```
// 0+ :deps
// 0+ items
// 1+ statement
```

`execute_evcxr` works by parsing "evcxr", building a binary crate from it in the temporary dir, executing the binary crate, and cleaning up.

Warning, because statements are supposed to be put in `fn main()`, macro invocations that expand to only items must have `#[expands_only_to_items]` attribute. Unfortunately, selective macro expansion is way too laborious. Even reasonably widely used `cargo expand` does not offer such a feature.

**Comment by JohnScience** [comment]
@ProfRon @davidlattimore

**Comment by davidlattimore** [comment]
@JohnScience - Nice! Evcxr pays a relatively high price in terms of complexity in order to support preserving variables between evaluations, so having a syntax-compatible alternative that just builds a binary is very welcome.

Some good news. Since you prompted me on this issue, I just retried what I described in [this comment](https://github.com/google/evcxr/issues/165#issuecomment-812850799) in rust-analyzer and it now correctly fills in the const generic argument in the type. So once we update to the latest rust-analyzer, the problem might actually be fixed in evcxr.

Unfortunately automatic publishes of rust-analyzer to crates.io has been broken for the last few months. However I spent some time earlier today (by coincidence) to see why it was failing and got a fix in - so assuming there are no other issues preventing automatic publishes, we'll hopefully have fresh rust-analyzer crates available some time next week. Then it will just be a matter of updating evcxr to use them - in three months, I'd expect there will have been some breaking changes.

**Comment by davidlattimore** [comment]
Just released evcxr 0.13.0 to crates.io. It includes updates to rust-analyzer, which as far as I can tell make const generics work. I note though that the rust-analyzer issue I linked above isn't closed out yet, so it's possible that there might still be some rough edges.
